src/flat.rs

```diff
@@ -0,0 +1,109 @@
+use std::marker::PhantomData;
+
+/// A flat buffer over a (multi channel) image.
+///
+/// Note that the strides need not conform to the assumption that constructed
+/// indices actually refer inside the underlying buffer but return values
+/// of library functions will always guarantee this. To manually make this
+/// check use `check_index_validities` and maybe put that inside an assert.
+pub struct FlatSamples<T, C: AsRef<[T]>> {
+    /// Underlying linear container holding sample values.
+    pub samples: C,
+
+    /// Add this to an index to get to the next sample in x-direction.
+    pub horizontal_stride: usize,
+
+    /// Add this to an index to get to the next sample in y-direction.
+    pub vertical_stride: usize,
+
+    /// Add this to an index to get to the sample in the next channel.
+    pub channel_stride: usize,
+
+    /// The width of the represented image.
+    pub width: u32,
```

Comment:
Should this be a u64 for greater generality? I know most of the library currently uses 32 bit ints for sizes, but given there are already checks before converting to ImageBuffer we don't necessary need to stick with that decision

Reply:
This is consistent with `GenericImageView::dimensions()` and guarantees that `width*height` does not overflow `usize` on 64-bit systems. Badly extrapolating current development in image capturing (x2 every five years), this has 17-18 more bits than currently required lasting some ~30-40 years.

Reply:
Fair enough. Much larger image sizes can definitely come up when working with other sorts of things like datasets but still probably won't run into the 32-bit size limit for quite some time. 

Do you know if there are any areas where it is common to store 1D data in image files? There it could be an issue much faster because a 4B x 1 image file would only be ~4GB which isn't hard to imagine. 

Regardless, I tend to err on the side of using 64-bits for everything unless have a strong reason not to, so it is probably fine just to ignore this concern.

Reply:
You are right, you might want to quickly reshape and still be able to represent all the intermediate formats. If we encourage that usage, then `u64` is better than `usize` because the compressed image representations from `stride = 0` make it possible to have images larger than memory, possibly *just* for extracting a subimage.